### PR TITLE
[WIP] [3.2] [Form] add 'empty_view' option in FormType

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17579, #18804 |
| License | MIT |
| Doc PR | todo |
- [x]  Gather feedback
- [x]  Use translation for default message
- [x]  Add some tests
- [x]  Update foundation 5 theme
- [ ] Make tests pass
- [ ]  Add missing translations
- [ ]  Doc PR

The `empty_view` option can be a `string`, or a `boolean`. By default it is `'No field'` when compound and `false` otherwise. It uses the `translation_domain` option for its translation.

`CollectionType` inherits it but this can be avoid be setting `empty_view` to `false`.

The `empty_view` is defined in `FormType::finishView()` and overriden in `ChoiceType::finishView()` to be defined when `choices` and `preferred_choices` view vars are empty.

In the templates, the old behaviour is kept explicitly set to `false`.

So 2 ways of using it:
- passing a translatable simple string to custom a message instead of an empty `div` or an empty `select` input :

``` php
// src/AppBundle/Form/Type/TenantType.php
// ...
$builder->add('cars', \Symfony\Bridge\Doctrine\Form\Type\EntityType::class, array(
    'class' => \AppBundle\Entity\Cars::class,
    'empty_view' => 'No cars available', // will be escaped
    // ...
));
```
- overriding the `empty_row` to use custom `html` :

``` twig
{# 'default/rent_a_car.html.twig #}
{# ... #}
{% form_theme _self %}
{% block 'empty_row' %}
    {# custom html with access to 'full_name', 'attr', and 'empty_view' view vars #}
{% endblock %}
{{ form(form) }}
```

What do you think ?

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 1]
```diff
@@ -330,6 +337,7 @@ public function configureOptions(OptionsResolver $resolver)
         $resolver->setNormalizer('choices_as_values', $choicesAsValuesNormalizer);
 
         $resolver->setAllowedTypes('choices', array('null', 'array', '\Traversable'));
+        $resolver->setAllowedTypes('empty_choices', array('null', 'string'));
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: On line 314 default for `'choices' => array(),` but line 339 allowed types is set to `array('null', 'array', '\Traversable')`.

Is `null` really needed ?

File: src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig [hunk 2]
```diff
@@ -165,21 +165,25 @@
 {%- endblock choice_widget_collapsed %}
 
 {% block choice_widget_expanded -%}
-    {% if '-inline' in label_attr.class|default('') %}
+    {% if '-inline' in label_attr.class|default('') -%}
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig]: looks like the whitespace control changes are unrelated to this feature and should be done in a different PR

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 3]
```diff
@@ -219,6 +219,12 @@ public function buildView(FormView $view, FormInterface $form, array $options)
             // POST request.
             $view->vars['full_name'] .= '[]';
         }
+
+        // Provide a default message when choices var is empty
+        if (empty($view->vars['choices']) && $options['expanded']) {
+            $view->vars['empty_choices'] = null !== $options['empty_choices']
+                ? $options['empty_choices'] : '<em>No choice available</em>';
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: Please do not mix messages and view data (which actually required you to use the `raw` filter in the form themes). Instead, you should only configure a message here and pass it to the `trans` filter in the form themes (by the way, you currently miss translating the message).

**Comment by xabbuh** [comment]
Status: Needs work

File: src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig (hunk 2 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig]: @xabbuh I did this to match bootstrap layout. DYTH that it needs a PR ?

Comment (xabbuh) [inline src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig]: Well, I think there's now use in having this different in 2.8/3.0.

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php (hunk 3 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: :+1:

I looked at the other components and only Security Core has its own translation domain `security`. For now form component has a validator catalogue with id's taking into account the ones of validator component.

Anyway I don't think that this message belongs to validator domain but the point is I wonder if I add a  `form.lang.xliff`, isn't this going to mess with apps when users have their own `form` domain ? Should't it be prefixed like 'sf_from.lang.xliff` ?

@xabbuh @aitboudad can you please light me on this ?

**Comment by HeahDude** [comment]
@xabbuh thanks for the review.

This PR is a POF to see if symfony deciders agree with this way of fixing #17579, I added your suggestion in the todo list of the description.

File: src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig (hunk 2 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig]: What do you mean by that ?

Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig]: @xabbuh I don't understand your comment.

**Comment by HeahDude** [comment]
Does the "feature" label added by __webmozart__ mean I should start fixing this PR ?

File: src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig (hunk 2 above)
Comment (jakzal) [inline src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig]: @HeahDude it's not related to this PR so should be done in a separate one. It should also be done on the latest branch this layout is available, to ease future merging.

Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/foundation_5_layout.html.twig]: @jakzal ok reverted. `foundation_5_layout` was introduced on 2.8 should I make a PR on this branch or on master ? Do you think it's really needed ?

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 4]
```diff
@@ -219,6 +219,12 @@ public function buildView(FormView $view, FormInterface $form, array $options)
             // POST request.
             $view->vars['full_name'] .= '[]';
         }
+
+        // Provide a default message when choices var is empty
+        if (empty($view->vars['choices']) && $options['expanded']) {
```
Comment (backbone87) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: I would remove this `if` completely. Its up to the template to decided what to display and when.
At least I would remove the check for `expanded` since one could decide to display a message insteadof an empty select box.

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: :+1: for removing the expanded condition.

But I don't think it's worth init a view var `empty_choices` and add it to the context if `choices` are not empty.

**Comment by HeahDude** [comment]
I've refactored a bit in the last commits and tests are green.

Status: Needs review

**Comment by HeahDude** [comment]
Tests are failing on travis but they are passing on my local environment :
```
Time: 6.86 seconds, Memory: 81.00Mb

OK (717 tests, 314 assertions)
OK ./src/Symfony/Bridge/Twig/Tests
```

Like in #17568, is something broken in travis, or tests are wrong ? ping @nicolas-grekas

**Comment by HeahDude** [comment]
Failure fix by @xabbuh in #17820

File: src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/choice_widget.html.php [hunk 5]
```diff
@@ -1,5 +1,17 @@
+<?php if (false !== isset($empty_choices)): ?>
 <?php if ($expanded): ?>
 <?php echo $view['form']->block($form, 'choice_widget_expanded') ?>
 <?php else: ?>
 <?php echo $view['form']->block($form, 'choice_widget_collapsed') ?>
 <?php endif ?>
+<?php else: ?>
+<?php if ('' === $empty_choices) {
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/choice_widget.html.php]: `$empty_choices` isn't set here (otherwise the block above would have been executed).

Comment (HeahDude) [inline src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/choice_widget.html.php]: right, the condition on the first line is wrong, it should be :
` +<?php if (!isset($empty_choices)): ?>`

Comment (HeahDude) [inline src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/choice_widget.html.php]: @xabbuh thanks, fixed.

**Comment by HeahDude** [comment]
I don't understand tests failure as they pass locally, expect for php templates of the `FrameworkBundle` I got this failure which I don't understand either :
```
7) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::testSingleChoiceExpandedWithoutChoices
Failed asserting that 

/div
    [not(@required)]
    [.="[trans]No choice available[/trans]"]
    [@id="my&id"]
    [@class="my&class"]

matches exactly once. Matches 0 times in 

<div id="my&amp;id" class="my&amp;class">
    [trans]No choice available[/trans]</div>

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/AbstractLayoutTest.php:84
/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/AbstractLayoutTest.php:106
/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/AbstractLayoutTest.php:970
```

Can someone please help me figure out what's wrong ?

**Comment by HeahDude** [comment]
Work is still in progress, but I've updated the description and refractor a bit in the last commit.

Status: Needs Review

**Comment by HeahDude** [comment]
I've got a suggestion as there is here an addition of a new default translation domain, may be it would be nice to translate the `'None'` default placeholder too (see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php#L283).

Do you agree with this handling ?
If so, may I add this in that PR (in a separated commit) or send another one if this one gets merged ?

File: src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0"?>
+<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
+    <file source-language="en" datatype="plaintext" original="file.ext">
+        <body>
+            <trans-unit id="1">
+                <source>No choice available</source>
+                <target><em>No choice available.</em></target>
```
Comment (backbone87) [inline src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf]: Either the <em> tag should be put in the correct xmlns or the whole content of <target> should be marked CDATA

Comment (HeahDude) [inline src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf]: @backbone87, thanks for the review, I've just copy-pasted `validators.en.xliff` as I'm not familiarized with this format, could you please precise ?

Comment (backbone87) [inline src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf]: its nothing specific to xliff. its about how xml namespaces work. when an xml parser parses the given xliff xml file, it interprets the em-element to be part of the xliff namespace (which is declared as the "default" namespace in the given document with `xmlns="urn:oasis:names:tc:xliff:document:1.2"`). This isnt a problem, unless the xliff consumer doesnt know what todo with an em-element. But xsd validation against the xliff schema will most likely fail (because it propably didnt define an em-element).

One way is to declare the xhtml namespace in the document ´xmlns:xh="http://www.w3.org/1999/xhtml"` and qualify the em-element against it: `<xh:em>...</xh:em>`. `http://www.w3.org/1999/xhtml` is used for xhtml 1.x and xml serialisation of html5 (xhtml5).

But since we dont care for the structure of the content in this use case, because the translation consumer expects `text/plain` produced by xliff, we can just properly escape the content of the target-element. either with `<target>&lt;em&gt;No choice available.$lt;/em$gt;</target>` or (more readable) `<target><![CDATA[<em>No choice available.</em>]]></target>` (cdata = character data)

Comment (HeahDude) [inline src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf]: @backbone87 Thank you very much for this explanation :)

**Comment by webmozart** [comment]
I think it's a good idea to add this functionality in general. When choices are loaded from a data source and no choice is available, displaying some text instead seems like the proper thing to do.

I'm not happy with the naming though. "empty", in the sense of a choice field, refers to  the "empty choice", i.e. the choice that is selected if *no choice* is selected. Using the same terminology creates ambiguity and potentially confuses our users. Unfortunately I can't think of a better name at the moment.

**Comment by HeahDude** [comment]
@webmozart, I proposed `empty_form` if this should be a global option for compound form types which would be inherited by `ChoiceType` (and `CollectionType`) in https://github.com/symfony/symfony/issues/17579#issuecomment-195442781.

We could add a default normalizer at the root definition for more safety to no use it when `compound` is `false`, but depending on the implementation it might just be ignored in that case.

**Comment by webmozart** [comment]
Even then `empty` is a problematic term. "Empty" - so far - refers to the *value* of a field. A field/form is empty if that value is `null` or `''`.

You are introducing a different concept of "empty", namely that a form is empty if doesn't contain children. Now it could happen that an empty form (your definition: no children) is not empty (existing definition: data is not `null`). Do you see how this can be ambiguous and confusing?

I think we need to find another name that does not involve the word "empty".

**Comment by HeahDude** [comment]
@webmozart, I've kept working in this, and did not use that idea but another one also with "empty"...

While implementing it globally, it felt more natural to use `empty_view` because actually it's the compound container block (HTML tag) which is empty and I found it makes a good echo to `empty_data` while keeping it clear that the value must be a string.

What about `no_child_view` or `empty_compound` ?

I just need to fix some tests with table layout and I will push everything.

**Comment by backbone87** [comment]
i rly like `empty_view`, but more acurate would be `on_empty_view` indicating that this string is only used **as** the view, if the *normally* rendered view is empty.

**Comment by HeahDude** [comment]
Ok I've rebased and made a clear diff between this old implementation and the new.

I kept the same tests for `ChoiceType` since it still has to override the parent's definition but had to adapt the `CollectionType` rendering tests. Is that a BC break ? If so it would be very minor.

I fail to make some tests pass, however I'm not an expert with Xpath, can anyone help me with this ?

**Comment by HeahDude** [comment]
Updated title and description accordingly.

File: src/Symfony/Component/Form/Resources/translations/sf_form.fr.xlf [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0"?>
+<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
+    <file source-language="en" datatype="plaintext" original="file.ext">
+        <body>
+            <trans-unit id="1">
+                <source>No choice available</source>
+                <target>![CDATA[<em>Aucun choix n'est disponible.</em>]]</target>
```
Comment (backbone87) [inline src/Symfony/Component/Form/Resources/translations/sf_form.fr.xlf]: the cdata tags have syntax errors. correct is: `<![CDATA[` and `]]>` (not escaped)

File: src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0"?>
+<xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
+    <file source-language="en" datatype="plaintext" original="file.ext">
+        <body>
+            <trans-unit id="1">
+                <source>No choice available</source>
+                <target>![CDATA[<em>No choice available.</em>]]</target>
```
Comment (backbone87) [inline src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf]: the cdata tags have syntax errors. correct is: `<![CDATA[` and `]]>` (not escaped)

Comment (backbone87) [inline src/Symfony/Component/Form/Resources/translations/sf_form.en.xlf]: actually i would move the em's to the template. i think that would be reasonable.

File: src/Symfony/Component/Form/Resources/translations/sf_form.fr.xlf (hunk 7 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Resources/translations/sf_form.fr.xlf]: @backbone87 thanks, I will update it.

**Comment by HeahDude** [comment]
I'm actually still stuck on that test:
```
1) Symfony\Bridge\Twig\Tests\Extension\FormExtensionDivLayoutTest::testEmptyCollection
Failed asserting that 

/div
    /em[.="[trans]No fields[/trans]"]
    [@id="my&id"]
    [@class="my&class"]
    [count(./div)=0]
/following-sibling::input[@type="hidden"][@id="names__token"]


matches exactly once. Matches 0 times in 

<div id="my&amp;id" class="my&amp;class">
        <em>[trans]No fields[/trans]</em>
    </div>
<input type="hidden" id="names__token" name="names[_token]">

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/AbstractLayoutTest.php:84
/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/AbstractLayoutTest.php:106
/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/AbstractDivLayoutTest.php:338
```

Does anyone know what I need to fix here ? :(

**Comment by HeahDude** [comment]
Rebased.

**Comment by HeahDude** [comment]
ping @webmozart :) There is no rush to merge any of my PR, we should polish anything but I can't go further for now without your review I guess.

Actually I've got three path tests failing on this one, and I can't make them pass...

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig [hunk 9]
```diff
@@ -232,6 +232,18 @@
...
 {%- endblock radio_row %}
 
+{# Support #}
+
+{% block empty_row -%}
+    {%- set translation_domain = translation_domain ?: 'sf_form' -%}
+    {%- set attr = attr|merge({class: (attr.class|default('') ~ ' form-control')|trim}) %}
+    <div {{ block('widget_container_attributes') }}>
+        <p>
+            <em>{{- translation_domain is same as(false) ? empty_view : empty_view|trans({}, translation_domain) -}}</em>
```
Comment (webmozart) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig]: Why `<em>`? I think we should refrain from adding styling as much as possible.

File: src/Symfony/Bridge/Twig/Resources/views/Form/form_table_layout.html.twig [hunk 10]
```diff
@@ -42,3 +42,14 @@
         {{- form_rest(form) -}}
     </table>
 {%- endblock form_widget_compound -%}
+
+{%- block empty_row -%}
+    {%- set translation_domain = translation_domain ?: 'sf_form' -%}
```
Comment (webmozart) [inline src/Symfony/Bridge/Twig/Resources/views/Form/form_table_layout.html.twig]: Why do you set this translation domain?

File: src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/empty_row.html.php [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<?php $translation_domain = $translation_domain ?: 'sf_form'; ?>
```
Comment (webmozart) [inline src/Symfony/Bundle/FrameworkBundle/Resources/views/Form/empty_row.html.php]: same here

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php [hunk 12]
```diff
@@ -244,6 +244,18 @@ public function finishView(FormView $view, FormInterface $form, array $options)
...
+
+        // We have to override parent {@link FormType} condition of the empty_view option here,
+        // even if there is no choices the field view may not be empty because:
+        //
+        //  - a select input is not "compound" but should be considered as any empty HTML tag,
+        //
+        //  - when expanded it could hold a placeholder which has been counted as a children,
+        //    we should ignore it in any case since its role is to be a "null" alternative to
+        //    choices, not a choice itself. Think of "Choose something" with no choices.
+        if (array() === $view->vars['choices'] && array() === $view->vars['preferred_choices']) {
```
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: Why is this check needed? Can't we decide in the view whether to output the text?

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig (hunk 9 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig]: OK

File: src/Symfony/Bridge/Twig/Resources/views/Form/form_table_layout.html.twig (hunk 10 above)
Comment (webmozart) [inline src/Symfony/Bridge/Twig/Resources/views/Form/form_table_layout.html.twig]: I see now. Shouldn't we rather call the catalog `forms.*.xlf` to be consistent with the other catalogs?

Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/form_table_layout.html.twig]: It was a suggestion that we might provide default translations for "No choices..", "No Fields" and even the actual "None" as default placeholder like it's done for validation message, there is already the dependency.

I can remove it of courses :)

File: src/Symfony/Component/Form/Extension/Core/Type/FormType.php [hunk 13]
```diff
@@ -157,6 +167,7 @@ public function configureOptions(OptionsResolver $resolver)
         $resolver->setDefaults(array(
             'data_class' => $dataClass,
             'empty_data' => $emptyData,
+            'empty_view' => $emptyView,
```
Comment (webmozart) [inline src/Symfony/Component/Form/Extension/Core/Type/FormType.php]: What about naming this `no_children_message`?

File: src/Symfony/Bridge/Twig/Resources/views/Form/form_table_layout.html.twig (hunk 10 above)
Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/form_table_layout.html.twig]: Yes I was just not sure it will not conflict with other catalogs with the same name in users code.

File: src/Symfony/Component/Form/Extension/Core/Type/FormType.php (hunk 13 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/FormType.php]: Sounds good :)

File: src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig (hunk 9 above)
Comment (backbone87) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig]: em isnt about styling? its about emphasis. and because there is no form, that one might expect, you put the emphasis on the presented message, probably giving reasons as for *why* there are no children.

Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig]: I agree, I did not even think about it when I added it naturally.

Because, it may be unexpected to see that string standing for a missing expectation.

The `emphasis` clearly mark the informational sense of such string IMHO.

Comment (webmozart) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig]: IMO we should leave that decision up to the developer. We never made such assumptions in any other part of form themes.

Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_3_layout.html.twig]: Fair enough then :)

File: src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php (hunk 12 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php]: @webmozart Because the view var `empty_view` is set [here](https://github.com/symfony/symfony/pull/17609/files/94f2a70c79494a4b248acf4cfb545757886dd059#diff-22a03074077455f3d1128486ef5d9851R118) but is not compatible with the only choice type.
I just thought we should resolve the state of this view var from within this method to avoid holding too much logic in the view.

I don't know how BC could be broken with such changes in form themes. the fact that I modify several test with the current implementation is not really reassuring. What do you think ?

**Comment by HeahDude** [comment]
Status: needs work

**Comment by nicolas-grekas** [comment]
ping @HeahDude, status of this PR considering feat. freeze?

**Comment by fabpot** [comment]
@HeahDude Do you think you can work on finishing this PR or shall we close it?

**Comment by fabpot** [comment]
Closing this PR as there is no more activity. @HeahDude Feel free to reopen whenever you have time to finish it. Thanks.

**Comment by lukepass** [comment]
Hello, sorry for reopening this one but I had a problem today with an empty CollectionType and I had to manually set setRendered in order to avoid the empty template. Is there another way to fix it?

Thanks!
